### PR TITLE
Flock patch 1

### PR DESCRIPTION
**Pull request by moonlol**
[ENHANCEMENT] 
## About the PR 
~~This PR is aimed to complete Flock to the point of it being a playable gamemode, expanding and fixing its current features while adding new content to it.~~
This PR is aimed to add QoL features and small additions aswell as bugfixes to Flock so that when an admin decides to spawn it in its a better expierence instead of a buggy mess.

Currently, I am working on it with @comradef191, @enakai, @ryebek & @PrimeNumb 

DONE:
Too many to list here, see spreadsheet [here](https://docs.google.com/spreadsheets/d/1mzGmDbZfEHFDc4kkYbSQAr8qCHnZ8JIIq7lr6jEgLRg/edit?usp=sharing)
TODO:
- touch up on a few more things
- add a flock gamemode
## Why's this needed? 
~~Flock was left in a partially completed state, it would be a waste to let it go unfinished. The resources & plans needed to complete it were already in place.~~
First patch of many. this contains many small additions or bugfixes for flock to make it a better expierence when an admin decides to spawn it in

## Changelog

```
(u)Moonlol & F191 & RichardGere & Rybek & Enakai & Pali:
(*)Added flock gamemode and antag setting. FLOCK IS NOT IN THE SECRET ROTATION.
(*)Flockminds now use a rift structure instead of a single egg when starting.
(*)Flock nanite spray harm intent added.
(+)Flockdrone ai will butcher dead drones.
(+)Added a number of flock examinations to flock items.
(+)Rift sprite by Enakai. Resource cube sprite by Pali.
```

### CODE REVIEW

**Comment by trerri** [comment]
woah nice

**Comment by ErwinSnowden** [comment]
I feel that instead of being a normal in-rotation antagonist, flockmind should be unlocked via easter egg much like the Solarium, with some necessary integration into the Flocktrader in some fashion. This would stop it from getting old and allow it to have its gimmickry without being a bore to the crew because there will always be other antagonists afoot with it.

**Comment by trerri** [comment]
@ErwinSnowden that sounds griefy. would be dope if it pulled from ghosts tho

**Comment by ErwinSnowden** [comment]
> @ErwinSnowden that sounds griefy. would be dope if it pulled from ghosts tho

The solarium ruins rounds by denying greentext and it's arguably griefy and extremely disruptive to the round, but nonantags can do it. It shouldn't change the meta at all and should have some degree of rarity.

**Comment by moonlol** [comment]
Ok first of all. green text means nothing.
Having flock be unlocked (as in just do it once and the antag is in normal rotation) by some event or puzzle sounds good. But having flock exclusively behind said puzzle is a terrible idea. Due to a number of factors.

1. people would get sick of flock, as the first week or w/e after the puzzle is solved it'll be just flock round after flock round after flock round.
2. People would use it to self antag.
3. People wont share the way to do it because "muh secrets"

i dont see why we cant have flock alongside other antags. for example in "mixed action" or even having 1-2 traitors in the flock gamemode

**Comment by PrimeNumb** [comment]
> I feel that instead of being a normal in-rotation antagonist, flockmind should be unlocked via easter egg much like the Solarium, with some necessary integration into the Flocktrader in some fashion. This would stop it from getting old and allow it to have its gimmickry without being a bore to the crew because there will always be other antagonists afoot with it.

An interesting idea regarding a flock puzzle, but I am not sure locking the gamemode behind it is a good idea. If the gamemode "gets old" as you put it then that either means the gamemode is played way too often or it isn't actually fun. Obviously if the latter is an issue then that should be tackled from a design perspective rather than shoving it in the "rare gamemode bin".

File: code/mob/living/critter/flock/flockdrone.dm [hunk 1]
```diff
@@ -707,23 +714,25 @@
 	if (user.floorrunning)
 		return // you'll need to be out of the floor to do anything
 	// CONVERT TURF
-	if(!isturf(target))
+//	message_admins("[target] target pre")//debugmessages
```
Comment (pali6) [inline code/mob/living/critter/flock/flockdrone.dm:717]:
```suggestion
```

File: code/mob/living/critter/flock/flockdrone.dm [hunk 2]
```diff
@@ -707,23 +714,25 @@
...
+	if(!isturf(target) && !(istype(target, /obj/storage/closet/flock) || istype(target, /obj/table/flock) || istype(target, /obj/structure/girder) || istype(target, /obj/machinery/door/feather)))
 		target = get_turf(target)
+//	message_admins("[target] target post")//debugmessages
```
Comment (pali6) [inline code/mob/living/critter/flock/flockdrone.dm:720]:
```suggestion
```

File: code/mob/living/critter/flock/flockdrone.dm [hunk 3]
```diff
@@ -707,23 +714,25 @@
...
-					actions.start(new/datum/action/bar/flock_convert(target, 10), user)
-			else if(user.a_intent == INTENT_DISARM)
+//			var/anything_needs_repair = flocktarget.broken
+//			for(var/obj/machinery/door/feather/D in flocktarget)
+//				anything_needs_repair = anything_needs_repair || D.broken
+//			if(anything_needs_repair)
+//				if(user.resources < 10)
+//					boutput(user, "<span class='alert'>Not enough resources to repair (you need 10).</span>")
+//				else
+//					actions.start(new/datum/action/bar/flock_convert(target, 10), user)
```
Comment (pali6) [inline code/mob/living/critter/flock/flockdrone.dm:734]: Should probably be removed? Or why is it commented out?

File: code/mob/living/critter/flockcritter_parent.dm [hunk 4]
```diff
@@ -415,3 +435,131 @@
...
+			pool(src.decal)
+		var/turf/T = get_turf(target)
+		playsound(T, "sound/impact_sounds/Glass_Shatter_3.ogg", 25, 1)
+		var/obj/item/raw_material/shard/S = unpool(/obj/item/raw_material/shard)
+		S.set_loc(T)
+		S.setMaterial(getMaterial("gnesisglass"))
+		S = unpool(/obj/item/raw_material/shard)
+		S.set_loc(T)
+		S.setMaterial(getMaterial("gnesis"))
+		qdel(target)
```
Comment (pali6) [inline code/mob/living/critter/flockcritter_parent.dm:565]: It's the end of an action so probably not necessary but it's good practice to do target = null after this.

File: code/modules/chemistry/Chemistry-Holder.dm [hunk 5]
```diff
@@ -59,6 +59,8 @@ datum
 
 		var/postfoam = 0 //attempt at killing infinite foam
 
+		var/gnesisstasis = 0//used for pausing gnesis for stuff
```
Comment (pali6) [inline code/modules/chemistry/Chemistry-Holder.dm:62]: I don't quite like this way of doing this. Couldn't it be a var on the gnesis reagent itself?

**Comment by pali6** [comment]
Oh and also in order to make ghost notifications work you need to change `if (!istype(C,/mob/dead/observer))` into `if (!istype(C.mob, /mob/dead/observer))` in the corresponding file.

**Comment by ZeWaka** [comment]
Changelog is a bit too wordy. Please compress.

File: code/obj/table.dm [hunk 6]
```diff
@@ -873,7 +866,7 @@
 
 	onUpdate()
 		..()
-		if (the_table == null || the_tool == null || owner == null || get_dist(owner, the_table) > 1)
+		if ((the_table == null || the_tool == null || owner == null || get_dist(owner, the_table) > 1) && !istype(owner, /mob/living/critter/flock/drone)) //allows flock drones to decon flock tables
```
Comment (ZeWaka) [inline code/obj/table.dm:869]: Can you explain this a bit more? Comment's not the most clear.

File: code/obj/flock/structure/rift.dm [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+/obj/flock_structure/rift/building_specific_info()
+	var/time_remaining = round(src.build_time - getTimeInSecondsSinceTime(src.time_started))
+	return "Approximately <span class='bold'>[time_remaining]</span> second[time_remaining == 1 ? "" : "s"] left until entry."
+
+/obj/flock_structure/rift/process()
+	var/elapsed = getTimeInSecondsSinceTime(src.time_started)
+	if(elapsed >= build_time)
+		src.visible_message("<span class='text-blue'>Multiple shapes exit out of [src]!</span>")
+//		new /mob/living/critter/flock/drone(get_turf(src), src.flock)
```
Comment (ZeWaka) [inline code/obj/flock/structure/rift.dm:31]: del
```suggestion
```

File: code/obj/flock/structure/rift.dm [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		var/j = pick(3, 4) + 1
+		for(var/i=1, i<j, i++)
+			var/obj/item/flockcache/x = new(src.contents)
+			x.resources = rand(40, 50)
+			eject += x
+		for(var/i=1, i<5, i++)
+			var/obj/flock_structure/egg/e = new(src.contents)
+			eject += e
+			e.flock = mainflock
```
Comment (ZeWaka) [inline code/obj/flock/structure/rift.dm:40]: Some explanation behind this would be nice.

File: code/obj/flock/goods.dm [hunk 9]
```diff
@@ -18,13 +18,18 @@
...
+	mat_changeappearance = 0
 
 /obj/item/reagent_containers/gnesis/New()
 	..()
 	var/datum/reagents/R = new/datum/reagents(50)
 	reagents = R
 	R.my_atom = src
 	R.add_reagent("flockdrone_fluid", 50)
+	src.setMaterial(getMaterial("gnesis"))
+	src.mat_changename = 0
```
Comment (ZeWaka) [inline code/obj/flock/goods.dm:32]: don't think this line is needed? it's added above.

File: code/mob/living/critter/flockcritter_parent.dm [hunk 10]
```diff
@@ -342,20 +342,40 @@
...
+				if(T)
+					F.tri_message("<span class='notice'>[owner] begins spraying glowing fibres onto [T].</span>",
+						F, "<span class='text-blue'>You begin repairing [T.real_name]. You will both need to stay still for this to work.</span>",
+						T, "<span class='text-blue'>[F.real_name] begins repairing you. You will both need to stay still for this to work.</span>",
+						"You hear hissing and spraying.")
+				else
+					F.tri_message("<span class='notice'>[owner] begins spraying glowing fibres onto [T].</span>",
+						F, "<span class='text-blue'>You begin repairing [T]. You will both need to stay still for this to work.</span>",
+						T, "<span class='text-blue'>[F.real_name] begins repairing you. You will both need to stay still for this to work.</span>",
+						"You hear hissing and spraying.")
```
Comment (ZeWaka) [inline code/mob/living/critter/flockcritter_parent.dm:358]: This doesn't make sense. If T doesn't exist, you can't begin repairing [T]

File: code/modules/admin/admin.dm [hunk 11]
```diff
@@ -728,7 +729,7 @@ var/global/noir = 0
 
 				var/list/valid_modes = list("secret","action","intrigue","random","traitor","meteor","extended","monkey",
 				"nuclear","blob","restructuring","wizard","revolution", "revolution_extended","malfunction",
-				"spy","gang","disaster","changeling","vampire","mixed","mixed_rp", "construction","conspiracy","spy_theft","battle_royale", "vampire","assday", "football")
+				"spy","gang","disaster","changeling","vampire","mixed","mixed_rp", "construction","conspiracy","spy_theft","battle_royale", "vampire","assday", "football", "flock")
```
Comment (ZeWaka) [inline code/modules/admin/admin.dm:732]: I don't think this is a good idea, especially as most of the game_mode checks are completely unfinished.

Comment (pali6) [reply code/modules/admin/admin.dm:732]: That only adds it to the list of modes admins can set not as a game mode playable under usual circumstances.

Comment (ZeWaka) [reply code/modules/admin/admin.dm:732]: Yeah, still. We've given people buttons that say PLEASE DO NOT PRESS THIS, and people still press it vOv.

Comment (moonlol) [reply code/modules/admin/admin.dm:732]: i could add a "Probably wont work well, press at own risk" to it.

File: code/obj/flock/structure/rift.dm (hunk 8 above)
Comment (moonlol) [reply code/obj/flock/structure/rift.dm]: so im using the flockdronegibs proc to handle throwing things around randomly. here im just creating the flockdrone eggs and resource caches before adding them to the eject list. the eject list being what gets thrown about

File: code/obj/table.dm (hunk 6 above)
Comment (moonlol) [reply code/obj/table.dm:869]: huh. cant exactly remember _why_ i did it. i think that has something to do with the action starting but ending onUpdate. ill check if this causes flockdrones to be able to walk away from table disassembly. if it doesnt ill update the comment to be more clear. if it does ill have to work out the problem

Comment (moonlol) [reply code/obj/table.dm:869]: yeah moving away causes the action to stop

File: code/obj/flock/structure/rift.dm (hunk 8 above)
Comment (ZeWaka) [reply code/obj/flock/structure/rift.dm]: can you put that in a comment?

File: code/obj/table.dm (hunk 6 above)
Comment (moonlol) [reply code/obj/table.dm]: Changed out the code to use a custom central flock_decon action instead of trying to reuse the tool interact action

**Comment by ZeWaka** [comment]
See my changelog comment above.

**Comment by moonlol** [comment]
compressed hopefully enough?
